### PR TITLE
Mentions in toots get linked to correct account.

### PR DESCRIPTION
**Pull request by StCyr**
Signed-off-by: Cyrille Bollu <cyrpub@bollu.be>

* Resolves: #585
* Target version: master

### CODE REVIEW

File: src/components/TimelinePost.vue [hunk 1]
```diff
@@ -102,20 +97,39 @@ export default {
...
+                        // Create links for mentions
+			if (typeof this.item.source !== 'undefined') {
+	                        let source = JSON.parse(this.item.source)
+        	                if (typeof source.tag !== 'undefined') {
+                	                source.tag.forEach(tag => {
+                        	                if (tag.type === 'Mention') {
+							let shortname = tag.name.match(/^@[^@]+/)
+                                        	        let patt = new RegExp(shortname[0], "g")
+                                                	message = message.replace(patt, function(matched) {
+	                                                        var a = '<a href="' + tag.href + '">' + matched + '</a>'
```
Comment (StCyr) [inline src/components/TimelinePost.vue]: maybe we don't want to use tag.href here but rather construct a local one using OC.generateUrl?

Comment (StCyr) [reply src/components/TimelinePost.vue]: I have updated this in https://github.com/nextcloud/social/pull/667/commits/b5d729bda8dd3b12722d86aae5134fabc5ce9246

**Comment by StCyr** [comment]
should be good now :-)

File: src/components/TimelinePost.vue (hunk 1 above)
Comment (daita) [reply src/components/TimelinePost.vue]: i guess both works !

**Comment by daita** [comment]
This is what happens when I send one single private message (from my mastodon) to 3 accounts:
- @cult@social.artificial-owl.com  (the account from the screenshot)
- @testing (myself on mastodon)
- @cult@test.artificial-owl.com (a 3rd party account)

Selection_096

**Comment by StCyr** [comment]
hey @daita 

Could you send me your toot 'source and content fields?

I can't figure out what has happened to your toots based on your screenshot; it's too much mangled.

Cyrille

**Comment by StCyr** [comment]
hey @daita,

your toot felt in a corner case: you've the same "short-mention" (@cult) twice.

I'll may find a fix in the front-end but the problem really comes from the back-end which translates your toot's content to "@cult  @testing @cult let's test !" while @cult refers either to @cult@social.artificial-owl.com or @cult@test.artificial-owl.com.

## Reproducing the bug ##

Here's a replay of https://github.com/nextcloud/social/blob/f6d0ff756bc4e73775083b46651f0af6e188acf1/lib/Model/ActivityPub/ACore.php#L532 with your toot's content:

```php
php > $content='<p><span class=\"h-card\"><a href=\"https://social.artificial-owl.com/index.php/apps/social/@cult\" class=\"u-url mention\">@<span>cult</span></a></span>  <span class=\"h-card\"><a href=\"https://mastodon.xyz/@testing\" class=\"u-url mention\">@<span>testing</span></a></span> <br /><span class=\"h-card\"><a href=\"https://test.artificial-owl.com/apps/social/@cult\" class=\"u-url mention\">@<span>cult</span></a></span> </p><p>let\'s test !</p>';
php > $new_content = strip_tags($content);
php > print html_entity_decode($new_content, ENT_QUOTES | ENT_HTML5);
@cult  @testing @cult let's test !
```

As you see your toot's content is replaced by a string containing `@cult` twice which causes the issue in the front-end:

```js
let patt = new RegExp("@cult", "g")
message = message.replace(patt, function(matched) {
    var a = '<a href="acme.org/' + matched + '">' + matched + '</a>'
    return a
})
patt = new RegExp("@testing", "g")
message = message.replace(patt, function(matched) {
    var a = '<a href="acme.org/' + matched + '">' + matched + '</a>'
    return a
})
patt = new RegExp("@cult", "g")
message = message.replace(patt, function(matched) {
    var a = '<a href="acme.org/' + matched + '">' + matched + '</a>'
    return a
})
13:26:52.511 "<a href=\"acme.org/<a href=\"acme.org/@cult\">@cult</a>\"><a href=\"acme.org/@cult\">@cult</a></a>  <a href=\"acme.org/@testing\">@testing</a> <a href=\"acme.org/<a href=\"acme.org/@cult\">@cult</a>\"><a href=\"acme.org/@cult\">@cult</a></a> let's test !"
```

## Solutions ##

A possible solution is to find a better regexp to avoid replacing mention in <a> tags.

But a better solution would be to fix the back end so that it replaces your toot's content by (for example) `@cult@social.artificial-owl.com  @testing @cult@test.artificial-owl.com let's test !`

**Comment by StCyr** [comment]
J'ai implémenté un parsing des hcards dans le backend. 

qu'est-ce-que tu penses de ça maintenant @daita?

**Comment by StCyr** [comment]
It seems that hcards are Mastodon's implementation of mentions.

Friendica displays them differently. see https://github.com/nextcloud/social/issues/328

**Comment by daita** [comment]
Seems superseded by #1035

**Pull request closed without merging**